src/NServiceBus.Transport.SQS/SnsPreparedMessageBatcher.cs

```diff
@@ -0,0 +1,64 @@
+#nullable enable
+
+namespace NServiceBus.Transport.SQS
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+
+    static class SnsPreparedMessageBatcher
+    {
+        public static IReadOnlyList<SnsBatchEntry> Batch(IEnumerable<SnsPreparedMessage> preparedMessages)
+        {
+            var allBatches = new List<SnsBatchEntry>();
+            var currentDestinationBatches = new Dictionary<string, SnsPreparedMessage>();
+
+            var groupByDestination = preparedMessages.GroupBy(m => m.Destination, StringComparer.Ordinal);
+            foreach (var group in groupByDestination)
+            {
+                SnsPreparedMessage? firstMessage = null;
+                var payloadSize = 0L;
+                foreach (var message in group)
+                {
+                    if (string.IsNullOrEmpty(message.Destination))
+                    {
+                        continue;
+                    }
+
+                    firstMessage ??= message;
+
+                    // Assumes the size was already calculated by the dispatcher
+                    var size = message.Size;
+                    payloadSize += size;
+
+                    if (payloadSize > TransportConstraints.MaximumMessageSize)
+                    {
+                        allBatches.Add(message.ToBatchRequest(currentDestinationBatches));
+                        currentDestinationBatches.Clear();
+                        payloadSize = size;
+                    }
+
+                    // we don't have to recheck payload size here because the support layer checks that a request can always fit 256 KB size limit
+                    // we can't take MessageId because batch request ID can only contain alphanumeric characters, hyphen and underscores, message id could be overloaded
+                    currentDestinationBatches.Add(Guid.NewGuid().ToString(), message);
+
+                    var currentCount = currentDestinationBatches.Count;
+                    if (currentCount != 0 && currentCount % TransportConstraints.MaximumItemsInBatch == 0)
```

Comment:
Just a curiosity question - how come you use the reminder operator instead of checking if currentCount == maxItems? Is it more efficient?

Reply:
Good question. Actually I don't know why I used this style. I'll simplify it